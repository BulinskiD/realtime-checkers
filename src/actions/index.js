import {TEST_ACTION} from '../constants/actionTypes';

export const testAction = () => {
    return {
        type: TEST_ACTION
    }
}
