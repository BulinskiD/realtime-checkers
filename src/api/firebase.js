import app from "firebase/app";
import "firebase/storage";
import "firebase/auth";
import "firebase/firestore";

var firebaseConfig = {
  apiKey: "AIzaSyD3uWma_iug5o_0w6EmchIXp4qd054HqcE",
  authDomain: "realtime-checkers.firebaseapp.com",
  databaseURL: "https://realtime-checkers.firebaseio.com",
  projectId: "realtime-checkers",
  storageBucket: "realtime-checkers.appspot.com",
  messagingSenderId: "100295366319",
  appId: "1:100295366319:web:ec1c182d83b660e8"
};

// Initialize Firebase
app.initializeApp(firebaseConfig);

export const auth = app.auth();
export const firestore = app.firestore();
