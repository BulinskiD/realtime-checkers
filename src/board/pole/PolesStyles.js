import styled from "styled-components";

export const PoleContainer = styled.div`
        border: solid 1px black;
        display: flex;
        justify-content: center;
        align-items: center;
    `;

export const ActivePlace = styled.div`
        background-color: red;
        cursor: pointer;
        width: 100%;
        height: 100%;
    `;