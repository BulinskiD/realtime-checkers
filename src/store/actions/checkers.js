import {SET_NEW_STATE, SELECT_CHECKER} from "../constants/actionTypes";

export const selectChecker = (col, row) => {
    return {
        type: SELECT_CHECKER,
        payload: {col, row}
    }
}

export const setNewGameState = (id, gameState) => {
    return {
        type: SET_NEW_STATE,
        payload: {id, gameState}
    }
}
