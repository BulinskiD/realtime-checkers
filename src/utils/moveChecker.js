export default (checkersPosition, from, to) => {

    let row = from.color === 'black' ? from.row - 1 : from.row + 1;


    if((from.col - to.col) === 2) {
        checkersPosition = checkersPosition.filter(item => {
            if((from.col - 1 === item.col) && (row === item.row)) {
                return item.color === from.color;
            }
            return true;
        });
    }

    if((from.col - to.col) === -2) {
        checkersPosition = checkersPosition.filter(item => {
            if((from.col + 1 === item.col) && (row === item.row)) {
                return item.color === from.color;
            }
            return true;
        });
    }



    checkersPosition[checkersPosition.indexOf(from)].col = to.col;
    checkersPosition[checkersPosition.indexOf(from)].row = to.row;
    checkersPosition[checkersPosition.indexOf(from)].selected = false;

    return checkersPosition;
}