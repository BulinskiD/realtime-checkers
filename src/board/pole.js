import React from 'react';
import {connect} from 'react-redux';
import Checker from './checker';
import PropTypes from 'prop-types';
import { PoleContainer, ActivePlace } from "./boardStyles";
import {selectChecker, setActivePoles} from "../store/actions/checkers";
import { firestore } from "../api/firebase";
import { getPole, checkNextStatus } from "../utils/utilFunctions";
import moveChecker from '../utils/moveChecker';

const Pole = props => {

    const {selected, col, row, color} = props.pole ? props.pole : {};

    const handleMove = async () => {
        const {checkersPosition, hasNextMove, selectedChecker} = moveChecker(props.nextMove,
                                                                             props.checkersPosition,
                                                                             props.selectedChecker,
                                                                          {col: props.col, row: props.row});
        const status = checkNextStatus(props.status, hasNextMove);
        try {
            await firestore.collection("games").doc(props.id)
                        .update({status, checkersPosition, selectedChecker, nextMove: hasNextMove });
        } catch (error) {
            console.log(error);
        }
    }

    return (
      <PoleContainer>
          {props.pole && <Checker selected={selected} col={col} row={row} color={color} status={props.status} />}
          {(!props.pole && props.active) && <ActivePlace onClick={handleMove} />}
      </PoleContainer>
    );
}

const mapStateToProps = ({currentGame} , ownProps) => {
    const {checkersPosition, activePoles, selectedChecker, id, status, nextMove} = currentGame;
    return {
        id: id,
        status: status,
        nextMove: nextMove,
        checkersPosition: checkersPosition,
        pole: getPole(ownProps.col, ownProps.row, checkersPosition),
        selectedChecker: selectedChecker,
        active: activePoles && activePoles.filter(item => item.col === ownProps.col && item.row ===ownProps.row).length === 1
    }
}

const checkersShape = PropTypes.shape({
    selected: PropTypes.bool,
    col: PropTypes.number,
    row: PropTypes.number,
    color: PropTypes.string});

Pole.propTypes = {
    id: PropTypes.string,
    status: PropTypes.string,
    nextMove: PropTypes.bool,
    checkersPosition: PropTypes.array,
    pole: checkersShape,
    active: PropTypes.bool,
    col: PropTypes.number,
    row: PropTypes.number
}

export default connect(mapStateToProps, {selectChecker, setActivePoles})(Pole);