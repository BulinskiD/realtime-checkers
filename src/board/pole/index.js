import React from 'react';
import {connect} from 'react-redux';
import styled from 'styled-components';
import Checker from '../checker';
import { firestore } from "../../api/firebase";
import getPole from '../../utils/getPole';
import moveChecker from '../../utils/moveChecker';

const Pole = props => {
    const PoleContainer = styled.div`
        border: solid 1px black;
        display: flex;
        justify-content: center;
        align-items: center;
    `;

    const ActivePlace = styled.div`
        background-color: red;
        cursor: pointer;
        width: 100%;
        height: 100%;
    `;

    const handleMove = async () => {
        try {
            await firestore.collection("games").doc(props.id).update({
                checkersPosition: moveChecker(props.checkersPosition, props.selectedChecker,  {col: props.col, row: props.row})
            });
        } catch (error) {
            console.log(error);
        }
    }

    return (
      <PoleContainer>
          {props.pole && props.pole.length !== 0 &&
          <Checker selected={props.pole[0].selected} col={props.pole[0].col} row={props.pole[0].row} color={props.pole[0].color} />}
          {props.active && <ActivePlace onClick={handleMove} />}
      </PoleContainer>
    );
}

const mapStateToProps = ({currentGame} , ownProps) => {
    const {checkersPosition, activePoles, selectedChecker, id} = currentGame;
    return {
        id: id,
        checkersPosition: checkersPosition,
        pole: getPole(ownProps.col, ownProps.row, checkersPosition),
        selectedChecker: selectedChecker,
        active: activePoles && activePoles.filter(item => item.col === ownProps.col && item.row ===ownProps.row).length === 1
    }
}

export default connect(mapStateToProps, null)(Pole);
