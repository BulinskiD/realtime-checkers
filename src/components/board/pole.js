import React from "react";
import { connect } from "react-redux";
import Checker from "./checker";
import PropTypes from "prop-types";
import { PoleContainer, ActivePlace } from "../../styles/boardStyles";
import { selectChecker, setActivePoles } from "../../store/actions/checkers";
import { firestore } from "../../api/firebase";
import {
  getPole,
  checkNextStatus,
  checkIfWinner
} from "../../utils/utilFunctions";
import moveChecker from "../../utils/moveChecker";
import handleError from "../../utils/handleError";

export const Pole = props => {
  const { selected, col, row, color, isKing } = props.pole ? props.pole : {};

  const handleMove = async () => {
    let oldFrom = {};
    // Save selectedChecker in oldForm
    Object.assign(oldFrom, props.selectedChecker);

    const { checkersPosition, hasNextMove, selectedChecker } = moveChecker(
      props.nextMove,
      props.checkersPosition,
      props.selectedChecker,
      { col: props.col, row: props.row }
    );
    let status = checkNextStatus(props.status, hasNextMove);

    // If its last move of color, save old from as empty object
    if (!hasNextMove) oldFrom = {};

    const winner = checkIfWinner(checkersPosition, props.status);
    if (winner) status = winner;

    try {
      await firestore
        .collection("games")
        .doc(props.id)
        .update({
          status,
          checkersPosition,
          selectedChecker,
          nextMove: hasNextMove,
          from: oldFrom,
          updated: Date.now()
        });
    } catch (error) {
      handleError(error);
    }
  };

  return (
    <PoleContainer col={props.col} row={props.row}>
      {props.pole && (
        <Checker
          isKing={isKing}
          isActiveTurn={props.isActiveTurn}
          selected={selected}
          col={col}
          row={row}
          color={color}
          status={props.status}
        />
      )}
      {!props.pole && props.active && (
        <ActivePlace
          className="active"
          onClick={() => props.isActiveTurn && handleMove()}
        />
      )}
    </PoleContainer>
  );
};

const mapStateToProps = ({ currentGame }, ownProps) => {
  const {
    checkersPosition,
    activePoles,
    selectedChecker,
    id,
    status,
    nextMove
  } = currentGame;
  return {
    id: id,
    status: status,
    nextMove: nextMove,
    checkersPosition: checkersPosition,
    pole: getPole(ownProps.col, ownProps.row, checkersPosition),
    selectedChecker: selectedChecker,
    active:
      activePoles &&
      activePoles.filter(
        item => item.col === ownProps.col && item.row === ownProps.row
      ).length === 1
  };
};

const checkersShape = PropTypes.shape({
  selected: PropTypes.bool,
  col: PropTypes.number,
  row: PropTypes.number,
  color: PropTypes.string
});

Pole.propTypes = {
  id: PropTypes.string,
  status: PropTypes.string,
  nextMove: PropTypes.bool,
  checkersPosition: PropTypes.array,
  pole: checkersShape,
  active: PropTypes.bool,
  col: PropTypes.number,
  row: PropTypes.number,
  isActiveTurn: PropTypes.bool,
  selectChecker: PropTypes.func,
  setActivePoles: PropTypes.func
};

export default connect(
  mapStateToProps,
  { selectChecker, setActivePoles }
)(Pole);
