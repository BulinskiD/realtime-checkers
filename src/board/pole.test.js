import React from 'react';
import {shallow} from 'enzyme';
import moveChecker from '../utils/moveChecker';
import {firestore} from '../api/firebase';
import { Pole } from './pole';
import {selectChecker, setActivePoles} from "../store/actions/checkers";
import { getPole, checkNextStatus } from "../utils/utilFunctions";
import handleError from "../utils/handleError";

jest.mock('../utils/handleError');
jest.mock('../utils/utilFunctions');
jest.mock('../api/firebase');
jest.mock('../utils/moveChecker');

const doc = jest.fn();
const update = jest.fn();
firestore.collection = jest.fn();
firestore.collection.mockReturnValue({doc});
doc.mockReturnValue({update});


describe('Pole', () => {
    it('should render black checker for given props', () => {
        const props = {id: '2', status: 'white', nextMove: true, checkersPosition: [], pole: {color: 'black'}, active: false, col: 1, row: 0 }
        const component = shallow(<Pole {...props} />);
        expect(component).toMatchSnapshot();
    });

    it('should render active pole for given props', () => {
        const props = {id: '2', status: 'white', nextMove: true, checkersPosition: [], pole: null, active: true, col: 1, row: 0 }
        const component = shallow(<Pole {...props} />);

        expect(component.find('Connect(Checker)').exists()).toBeFalsy();
        expect(component.find('.active').exists()).toBeTruthy();
        expect(component).toMatchSnapshot();
    });

    it('should handle move properly', async () => {
        const selectChecker = jest.fn();
        const setActivePoles = jest.fn();
        const props = {id: '2', status: 'white', nextMove: true, checkersPosition: [], pole: null, active: true, col: 1, row: 0,
            selectChecker, setActivePoles};
        moveChecker.mockReturnValue({checkersPosition: [], hasNextMove: false, selectedChecker: {}});
        checkNextStatus.mockReturnValue('black');
        const prom = Promise.resolve('OK');
        update.mockReturnValue(prom);

        const component = shallow(<Pole {...props} />);

        component.find('.active').simulate('click');

        await prom;
        expect(update).toBeCalledWith({status:'black', checkersPosition: [], selectedChecker: {}, nextMove: false});
    });

    it('should call handle error when request fails', async () => {
        expect.assertions(2);

        const selectChecker = jest.fn();
        const setActivePoles = jest.fn();
        const props = {id: '2', status: 'white', nextMove: true, checkersPosition: [], pole: null, active: true, col: 1, row: 0,
            selectChecker, setActivePoles};
        moveChecker.mockReturnValue({checkersPosition: [], hasNextMove: false, selectedChecker: {}});
        checkNextStatus.mockReturnValue('black');
        const prom = Promise.reject('Error');
        update.mockReturnValue(prom);

        const component = shallow(<Pole {...props} />);

        component.find('.active').simulate('click');
        expect(update).toBeCalledWith({status:'black', checkersPosition: [], selectedChecker: {}, nextMove: false});
        try {
            await prom;
        } catch(error) {
            expect(handleError).toBeCalledWith('Error');
        }
    });
});