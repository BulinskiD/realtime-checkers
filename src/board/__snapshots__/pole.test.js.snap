// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pole should render active pole for given props 1`] = `
<styled.div>
  <styled.div
    className="active"
    onClick={[Function]}
  />
</styled.div>
`;

exports[`Pole should render black checker for given props 1`] = `
<styled.div>
  <Connect(Checker)
    color="black"
    status="white"
  />
</styled.div>
`;
